File: core/Sources/ChotkiCore/Platform/Notifier.swift
import Foundation

/// One button on a notification.
public struct NotificationAction: Sendable, Hashable, Codable {
    public let id: String
    public let title: String

    public init(id: String, title: String) {
        self.id = id
        self.title = title
    }

    /// Copy is deliberately neutral. See the Tone section in design.md:
    /// a reminder says what is due, never how overdue it is.
    public static let markComplete = NotificationAction(id: "complete", title: "Mark complete")
    public static let snooze = NotificationAction(id: "snooze", title: "Snooze an hour")
}

public struct NotificationRequest: Sendable, Hashable, Codable {
    /// Stable, so the scheduler can cancel every pending reminder for one occurrence.
    public let id: String
    public let title: String
    public let body: String
    public let actions: [NotificationAction]

    public init(id: String, title: String, body: String, actions: [NotificationAction] = []) {
        self.id = id
        self.title = title
        self.body = body
        self.actions = actions
    }
}

public struct NotificationActionEvent: Sendable, Hashable {
    public let requestID: String
    public let actionID: String

    public init(requestID: String, actionID: String) {
        self.requestID = requestID
        self.actionID = actionID
    }
}

/// Implemented per platform. Core decides *when*; this only *shows*.
///
/// `supportsActions` is not decoration — several Linux notification daemons
/// ignore action buttons entirely, and the reminder must degrade to a plain
/// notification rather than being lost.
public protocol Notifier: Sendable {
    var supportsActions: Bool { get }
    func requestAuthorization() async throws -> Bool
    func show(_ request: NotificationRequest) async throws
    func cancel(ids: [String]) async
    var actionEvents: AsyncStream<NotificationActionEvent> { get }
}

File: android/core/src/main/kotlin/org/chotki/core/Notifier.kt
package org.chotki.core

import kotlinx.serialization.Serializable

/** One button on a notification. */
@Serializable
data class NotificationAction(val id: String, val title: String) {
    companion object {
        /**
         * Copy is deliberately neutral: a reminder says what is due, never how
         * overdue it is.
         */
        val MARK_COMPLETE = NotificationAction("complete", "Mark complete")
        val SNOOZE = NotificationAction("snooze", "Snooze an hour")
    }
}

@Serializable
data class NotificationRequest(
    /** Stable, so the scheduler can cancel every pending reminder for one occurrence. */
    val id: String,
    val title: String,
    val body: String,
    val actions: List<NotificationAction> = emptyList(),
)

data class NotificationActionEvent(val requestID: String, val actionID: String)

/**
 * Implemented per platform. Core decides *when*; this only *shows*.
 *
 * `supportsActions` is not decoration — a platform that ignores action buttons
 * must degrade to a plain notification rather than losing the reminder.
 *
 * Synchronous, unlike the Swift original, which returns an `AsyncStream` of
 * action events. Adding a coroutines dependency to `:core` for one callback is
 * not a trade worth making: Android pushes action events in rather than core
 * pulling them out.
 */
interface Notifier {
    val supportsActions: Boolean
    fun requestAuthorization(): Boolean
    fun show(request: NotificationRequest)
    fun cancel(ids: List<String>)
}
